software/controller/lib/core/commfail_alarm.cpp

```diff
@@ -0,0 +1,69 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "commfail_alarm.h"
+
+// There should be a packet from GUI for every 30ms. If a packet is not
+// recived periodically, then assuming something went wrong we wait for
+// 100ms before raising a communication failure alarm
+static constexpr Duration kCommunicationTimeoutMs{milliseconds(100)};
+
+// This count checks how many times alarm is triggered before it raises alarm
+// Alarm will be raised immediately considering as high priority alarm
+static constexpr uint8_t kTriggerCountThreshold{1};
+
+// Initialize the time_stamp vartiable for CommFail_Alarm object
+void CommFailAlarm::Initialize(Time ts) {
+  time_stamp_ = ts;
+  debug.Print("Communication Failure Alarm Initialized\n");
+}
+
+// This function checks for communication timeout, if yes raises alarm, if
+// no suppresses alarm. in both the cases it notes the occurance time stamp
+bool CommFailAlarm::Handler(Time ts, Time lastrx) {
+  // does the communication timed out
+  if (ts - lastrx > kCommunicationTimeoutMs) {
+    CommunicationFailed(ts);
+  } else {
+    CommunicationResumed(ts);
+  }
+  return triggered_;
+}
+
+// communication failed so raise alarm
+void CommFailAlarm::CommunicationFailed(Time ts) {
+  if (trigger_count_ < kTriggerCountThreshold) {
+    trigger_count_++;
+    if (trigger_count_ == kTriggerCountThreshold) {
+      // take the time stamp when alarm occured
+      time_stamp_ = ts;
+      triggered_ = true;
+      debug.Print("Communication Failed, Raised Alarm\n");
+    }
+  }
+}
+
+// communication resumed so clear alarm
+void CommFailAlarm::CommunicationResumed(Time ts) {
+  if (trigger_count_ > 0) {
+    trigger_count_--;
```

Comment:
So, if communication failed 3 times, but then communication is resumed, it will need to be resumed 3 times before the alarm is disabled? Please explain this to me.

Reply:
1. One way it will filter glitches, if we increase threshold count in future.
2. The thought holds good in clearing all the number of times error got triggered for a queue of occurrences. a present the threshold is just one.

Reply:
> If it went 1 minute without communications and now it is communicating correctly again, you are saying it will have to wait for another minute to clear an equal number of triggers before it thinks that everything is finally ok?

Not as I understand it... trigger_count_ stops being increased when the threshold is reached. meaning so long as the threshold is one, the alarm is suppressed as soon as communication resumes.

However, I have to say it feels weird to me that if we were to increase kTriggerCountThreshold (to filter out glitches), then the alarm wouldn't stop as soon as communications resumes. A proper communication is by definition not a glitch, and shouldn't be filtered out.